### PR TITLE
stdenv/generic: introduce shellDryRun

### LINKED ISSUE #126344: writeShellScript rejects valid scripts that use extglob
**Describe the bug**

`writeShellScript` and `writeShellScriptBin`'s check phase runs the script with `${stdenv.shell} -n` which ignores all commands, including `shopt`, so any script that sets and uses extglob (an option that alters bash's syntax) is rejected, even if valid.

**To Reproduce**

1. Try to build the following derivation:

```nix
pkgs.writeShellScript "foo" ''
  shopt -s extglob
  echo @(foo|bar)
''
```

2. Get this error:

```
error: builder for '/nix/store/gqpa7fk8v7fr5zzl44l1brc0gc408ks0-foo.drv' failed with exit code 2;
       last 2 log lines:
       > /nix/store/d6yjx34nqr4x970dcy9rrnajsdvy9mp3-foo: line 3: syntax error near unexpected token `('
       > /nix/store/d6yjx34nqr4x970dcy9rrnajsdvy9mp3-foo: line 3: `echo @(foo|bar)'
```

**Expected behavior**

This is a valid bash script, so it probably shouldn't be rejected.

**Additional context**

I'm running into this problem because Home Manager's [bash module](https://github.com/nix-community/home-manager/blob/aa36e2d6b481a54b95fbddbecb76076e0f38eb89/modules/programs/bash.nix#L171) uses `writeShellScript` to write the user's initialisation files, and I happen to set and use extglob in my `.bashrc`.

Of course, an immediate solution would be to stop using `writeShellScript` there, but I'm wondering if this problem could arise in other situations.

**Notify maintainers**

It seems like @grwlf added `writeShellScriptBin` and then @deliciouslytyped added `writeShellScript`.
Also pinging @rycee

### PR DESCRIPTION
**Pull request by ncfavier**
Add `shellDryRun` to the generic stdenv and substitute it for uses of `${stdenv.shell} -n`. The point of this layer of abstraction is to add the flag `-O extglob`, which resolves #126344 in a more direct way.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ncfavier** [comment]
@ofborg build tests.trivial-builders.overriding

File: pkgs/build-support/trivial-builders.nix [hunk 1]
```diff
@@ -121,18 +121,18 @@ rec {
         allowSubstitutes = false;
       }
       ''
-        n=$out${destination}
-        mkdir -p "$(dirname "$n")"
+        target=$out${destination}
```
Comment (happysalada) [inline pkgs/build-support/trivial-builders.nix:124]: ❤️

**Comment by happysalada** [comment]
I really like the naming changes.
I've tried to look for reasons to not have extglob enabled but couldn't find any. I think the only downside is that people might not know about it. Perhaps we should just add something in the release notes ? (no worries if you think it's not worth it).

One other thing that jumped to mind is that probably failglob should be enabled as well. I understand that this would be breaking behavior so not for this PR, just a thought that came to mind.

So all in all, looks good to me.

Thank you for this!

**Comment by ncfavier** [comment]
> I've tried to look for reasons to not have extglob enabled but couldn't find any. I think the only downside is that people might not know about it. Perhaps we should just add something in the release notes ? (no worries if you think it's not worth it).

The reason you might want extglob disabled in general is if you plan to run arbitrary `sh` scripts that might use extglob syntax without expecting it to glob. This is not the case at all for `writeShellScript`, where the scripts are statically known and can just be written as bash.

I don't think it's worth mentioning, but I won't stop people from writing release notes.

> One other thing that jumped to mind is that probably failglob should be enabled as well. I understand that this would be breaking behavior so not for this PR, just a thought that came to mind.

Do you mean enabled at runtime (since it wouldn't make a difference at check time)? If so, definitely not in `writeShellScript` (I think we want to keep these functions as agnostic as possible to the kind of shell script they accept), but maybe in the default builder or something?

I don't know if failglob is that useful in practice, given that most globs are passed to commands and most commands fail on non-existent files.

> So all in all, looks good to me.
> 
> Thank you for this!

Thanks for the review!

**Comment by happysalada** [comment]
I didn't think about the checktime vs runtime difference for failglob, it might not be that useful.

Regarding failglob discussion. I'm trying to enable failglob for stdenv, and there are 22 failures spotted so far, and possibly quite a bit left https://github.com/NixOS/nixpkgs/pull/142860 . Hopefully this is in by default sometimes in the future.

**Comment by happysalada** [comment]
I don't think it's worth delaying this anymore, let me know if you had intended to add something else.

**Comment by ncfavier** [comment]
This is ready. The darwin failures are unrelated, let's try restarting the build
@ofborg build tests.trivial-builders.overriding

**Comment by ncfavier** [comment]
Oh well.

**Comment by happysalada** [comment]
It looks like a time-out. I think it's fine to merge. Let's just keep an eye on the staging-next branch

**Comment by ncfavier** [comment]
Thanks!

**Pull request merged**